src/main/scala/com/twitter/finagle/websocket/WebSocketHandler.scala

```diff
@@ -1,25 +1,31 @@
 package com.twitter.finagle.websocket
 
-import com.twitter.concurrent.{Broker, Offer}
-import com.twitter.finagle.{CancelledRequestException, ChannelException}
-import com.twitter.finagle.util.DefaultTimer
-import com.twitter.util.{Future, Promise, Return, Throw, TimerTask, Try}
 import java.net.URI
 
-import org.jboss.netty.buffer.ChannelBuffers
-import org.jboss.netty.channel._
-import org.jboss.netty.handler.codec.http.websocketx._
-import org.jboss.netty.handler.codec.http.{HttpHeaders, HttpRequest, HttpResponse}
-import org.jboss.netty.handler.timeout.{IdleStateAwareChannelHandler, IdleStateEvent}
-
-import scala.collection.JavaConversions._
-
-class WebSocketHandler extends IdleStateAwareChannelHandler {
+import com.twitter.concurrent.{Broker, Offer}
+import com.twitter.finagle.CancelledRequestException
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.util._
+import io.netty.buffer.{ByteBuf, Unpooled}
+import io.netty.channel._
+import io.netty.handler.codec.http._
+import io.netty.handler.codec.http.websocketx._
+import io.netty.handler.timeout.IdleStateEvent
+import scala.collection.JavaConverters._
+
+class WebSocketHandler extends ChannelDuplexHandler {
   protected[this] val messagesBroker = new Broker[String]
   protected[this] val binaryMessagesBroker = new Broker[Array[Byte]]
   protected[this] val pingsBroker = new Broker[Array[Byte]]
   protected[this] val closer = new Promise[Unit]
-  protected[this] val timer = DefaultTimer
+  protected[this] val timer = DefaultTimer.twitter
+
+  protected def toByteArray(buffer: ByteBuf): Array[Byte] = {
```

Comment:
Can we use the backing array of ByteBuf directly instead of allocating a new buffer? See the example here: https://stackoverflow.com/questions/19296386/netty-java-getting-data-from-bytebuf

Reply:
The reason is WebSocket frame objects (e.g PingWebSocketFrame, BinaryWebSocketFrame) are not `Heap Buffers`. So, the data is not stored in JVM Heap space and cannot simply read as byte array.
To make the network transfer efficient, WebSocket uses `Direct Buffer` approach. 
Anyway we can remove my `toByteArray` and use Netty4 util function `ByteBufUtil.getBytes()`
I checked, the implementation.
It checked the buffer type and apply the correct conversion.